Disable cell-segmentation of import paths with buckconfig

cormacrelf opened the pull request:
Fixes https://github.com/facebook/buck2/issues/683 if you enable the config.

Also fixes https://github.com/facebook/buck2/issues/890.

This is achieved by ignoring the build_file_cell during ImportPath construction. I can confirm this fixes the test case on that issue (https://github.com/cormacrelf/buck2-tset-bug).

This is a big improvement for users outside of Meta because previously it was practically impossible to use transitive sets. I have been writing some rules and have had to emulate the feature with regular starlark sets.

Previous version of this PR had cfg flags; now I made a buckconfig for it. Because it does break some things.

This will break any code that reads `read_config` / `get_cell_name` at the top level of a BZL file outside of the prelude. It will also break code that uses `read_config` / `get_cell_name` during analysis (which should never have worked). It is therefore not activated for anyone by default. However it is not super difficult to migrate. To test it, you can use

```ini
# root .buckconfig
[buck2]
    disable_cell_segmentation = true
```

## Breakage

### Every user, with or without the buckconfig

#### `read_config` during analysis is no longer accepted

Previously this was unofficially and perhaps accidentally supported:

```starlark
def _impl(ctx: AnalysisContext) -> list[Provider]:
    xyz = read_config("abc", "xyz")
    ...

rule = rule(impl = _impl)
```

The reason being that `read_config` was marked as `#[starlark(speculative_exec_safe)]`, so it was actually being inlined when the file was initially parsed. It can no longer be inlined so this no longer works. I recall this working sometimes but not always, to my great puzzlement, which is why I expect people won't have used it that much. There will nevertheless be code that does rely on it. It now fails with

<details>
<summary>
<blockquote>
error: This function is unavailable during analysis (usual solution is to place the information on a toolchain)
</blockquote>
</summary>

```
BUILD FAILED
Error running analysis for `root//rust:test (prelude//platforms:default#7171795d350c5b11)`

Caused by:
    Traceback (most recent call last):
      File <builtin>, in <module>
      * prelude/rust/rust_binary.bzl:519, in rust_test_impl
          re_executor, executor_overrides = get_re_executors_from_props(ctx)
      * prelude/tests/re_utils.bzl:68, in get_re_executors_from_props
          re_arg = _get_re_arg(ctx)
      * prelude/tests/re_utils.bzl:20, in _get_re_arg
          force_local = read_config("fbcode", "disable_re_tests", default = False)

    error: This function is unavailable during analysis (usual solution is to place the information on a toolchain)
      --> prelude/tests/re_utils.bzl:20:19
       |
    20 |     force_local = read_config("fbcode", "disable_re_tests", default = False)
       |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       |
```

</details>


#### To migrate code that relied on starlark inlining of read_config

Simply lift `read_config` calls to constants outside of analysis impls. I have already done this for a few stragglers in the prelude.


```diff
+_XYZ = read_config("abc", "xyz")
+
def _impl(ctx: AnalysisContext) -> list[Provider]:
-    xyz = read_config("abc", "xyz")
-    if xyz:
+    if XYZ:

abc_library = rule(impl = _impl)
```

### Breaking changes WITH the buckconfig

#### Before (non-prelude cell)

```starlark
# othercell//:defs.bzl

# all calls relate to the build file cell and will agree.
STATIC_CONFIG = read_config(...)
STATIC_CELL = native.get_cell_name()
abc = rule(
    ...
    attrs = { "xyz": attrs.string(default = read_config(...)), },
)
def macro():
    _ = read_config(...)
    _ = native.get_cell_name()
```

```starlark
# BUCK file

load("@othercell//:defs.bzl", "macro")
macro()
```

#### After

There is now a difference between two kinds of calls to get_cell_name/read_config. It matters when you evaluate them.

```starlark
# othercell//:defs.bzl

# These now relate to the .bzl file's cell
STATIC_CONFIG = read_config(...)
STATIC_CELL = native.get_cell_name()
abc = rule(
    ...
    attrs = { "xyz": attrs.string(default = read_config(...)), },
)

def macro():
    # these two relate to build file cell as before
    _ = read_config(...)
    _ = native.get_cell_name()

```

```starlark
# BUCK file

load("@othercell//:defs.bzl", "macro")
macro()
```

#### For the prelude specifically

No change, the "after" is already how it works in the prelude.

#### How to migrate non-prelude cells

Before `buck2.disable_cell_segmentation`, **all** `read_config`/`get_cell_name` calls referred to the importing build file's cell. Achieving this is now a bit more difficult. You must now call `read_config` in a macro such that it is evaluated in the build file context. **This may require wrapping a rule in a macro.**

The ONLY way to achieve 'read the importing build file cell config' is to put it in a macro and call that macro during build file evaluation. Notably, **`attrs.string(default = read_config(...))` is now evaluated in the cell owning the .bzl file** since that is eagerly evaluated and passed to `rule()`.

```diff
def _impl(ctx: AnalysisContext) -> list[Provider]:
-    xyz = read_config("abc", "xyz")
-    if xyz:
+    if ctx.attrs.xyz:

-abc_library = rule(impl = _impl)
+abc_library = rule(impl = _impl, attrs = { "xyz": attrs.string() })
+_abc_library = abc_librneary
+
+def abc_library(**kwargs):
+    kwargs.setdefault("xyz", read_config("abc", "xyz"))
+    _abc_library(**kwargs)
```

I believe this pattern is already necessary to read package values and populate defaults with them.

Alternatively, use a toolchain rule and allow your users to configure things there instead of in buckconfig.

## Review questions

- ~~Is `parser.cell_segmentation` the best buckconfig name? Almost every other config is in the `buck2` section but this is one of those rare occasions that maybe parser is right? Feel free to change it in `app/buck2_interpreter/src/import_paths.rs`. You can call the config whatever you like, just fix the tests, or I am happy to change it. I suppose it doesn't matter since it's a temporary thing anyway.~~ Ok I changed it to `buck2.disable_cell_segmentation`.
- I don't know how to run the integration test that I wrote. Fingers crossed it works.
- There may be a need for disabling this in a more fine-grained way. Especially if people are using a bunch of external cells and they have not been updated to support this new mode.
- Do we maybe want to read the buckconfig not from the root, but rather from the bzl's cell? That way a cell can mark itself safe for single-eval. Which it will do if it contains transitive set definitions, for example. This would slow down a global migration away from cell segmentation by reducing incentive to change code to work without it.

cormacrelf committed:
Add new buckconfig `buck2.disable_cell_segmentation`

ImportPath::new_with_build_file_cells takes a flag `cell_segmentation`
and suppresses the effect of build_file_cell if it is false.
We then plumb this boolean through buckconfig and dice to all the
locations it is needed to construct an ImportPath.

Ultimately if `buck2.disable_cell_segmentation` is true, then ImportPath
becomes a redundant wrapper around CellPath.

cormacrelf committed:
LegacyBuckConfig::view() to get a view

This is needed sometimes in tests. You can just use `(&self.root_config).xxx()` but this is
hard to discover. We could also implement the LegacyBuckConfigView directly on LegacyBuckConfig
but you generally want to prohibit mutating it, hence implementing on &T in the first place.

cormacrelf committed:
Privatize ImportPath::testing_new_cross_cell

This helps avoid having to add more cell_segmentation parameters by reducing
the public API of ImportPath.

cormacrelf committed:
Add test for cell segmentation path resolution

Needed to change the original test when experimenting
with making disable_cell_segmentation the default on OSS.
So may as well test the actual cell_segmentation behaviour.

cormacrelf committed:
Add an integration test for cell segmentation

cormacrelf committed:
oss shim: stop trying to read_config/get_current_cell at top level

This lets buck build itself again. These were top level get_cell_name/read_config calls
that can simply be moved into the macro call.

cormacrelf committed:
read_config is no longer speculative_exec_safe

The comment explains why. This will cause some breakage because there is code in prelude
relying on read_config being inlined. Fortunately we have this behind a buckconfig flag.

cormacrelf committed:
Fix read_config during analysis for rust_test

Previously worked because of #[starlark(speculative_exec_safe)]. Move to top level instead
of calling during analysis.

cormacrelf committed:
More documentation on get_cell_name and read_config

ahornby commented on tests/core/interpreter/test_cell_segmentation_data/one/rules.bzl:
```diff
@@ -0,0 +1,54 @@
+Info = provider(
```
```suggestion
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is dual-licensed under either the MIT license found in the
# LICENSE-MIT file in the root directory of this source tree or the Apache
# License, Version 2.0 found in the LICENSE-APACHE file in the root directory
# of this source tree. You may select, at your option, one of the
# above-listed licenses.

Info = provider(
```

cormacrelf commented:
> Before buck2.disable_cell_segmentation, all read_config/get_cell_name calls referred to the importing build file's cell. Achieving this is now a bit more difficult. You must now call read_config in a macro such that it is evaluated in the build file context. This may require wrapping a rule in a macro.

I think we need `read_config(buildfile_cell=True)` that can run at analysis time. Unless, this being 6 months later, I have forgotten too much about rule evaluation and DICE and this won't be possible.

Macros do work but they wreck the documentation and we should be making it so standalone rules have enough functionality that macros are not required.

8Keep commented:
@cormacrelf would you be able to rebase this one?

cormacrelf committed:
Fix some breakage in e2e

cormacrelf commented:
Rebased. I am looking into breakage in the e2e test suite, at least the ones I have been able to run with https://github.com/facebook/buck2/pull/1311. But you will probably get much cleaner results inside meta, many tests are broken in oss.

I already fixed some prelude things in the PR and the test I added. The only e2e regression I found so far seems to be the spin loop test harness code in test_cancellation_data. Looks fairly simple to fix.

I came up with a concept to help with the breakage: `read_owner_config`. It covers most of the uses of read_config that were broken. Clearly people were using (and misusing) read_config in useful ways that we probably still need and probably don't want to wrap stuff in macros for. I will PR separately if you are interested, I added some tests etc, or you can just follow through on what it implies.

<details>
<summary>read_owner_config</summary>

```rust
    /// Like `read_config`, but always reads from the `.buckconfig` of the cell that
    /// *owns the target*, regardless of where the reading code lives or where the
    /// chain of `load()`s started.
    ///
    /// This is a more predictable alternative to `read_config` for rule authors:
    /// its result does not depend on cell segmentation (see `read_config`), and it
    /// behaves identically whether you call it eagerly in a macro or lazily during
    /// analysis.
    ///
    /// * During `BUCK` file evaluation (e.g. inside a macro), it reads the build
    ///   file's cell, which is the cell that owns the targets being declared. In this
    ///   context it is exactly equivalent to `read_config`.
    /// * During analysis of a target, it reads the cell that owns the target being
    ///   analysed (i.e. `ctx.label.cell`).
    /// * At the top level of a `.bzl` file there is no owning target, so this errors.
    ///   In particular, `attrs.string(default = read_owner_config(...))` is not
    ///   allowed, because the default is evaluated eagerly when `rule()` is
    ///   constructed; move the call into the rule `impl` or a macro instead.
    ///
    /// ```python
    /// # cell//rules.bzl
    /// def _impl(ctx: AnalysisContext) -> list[Provider]:
    ///     # reads the config of whichever cell owns `ctx.label`
    ///     flag = read_owner_config("my_rules", "flag", default = "off")
    ///     ...
    /// my_rule = rule(impl = _impl, attrs = {})
    /// ```
    //
    // Not `speculative_exec_safe`: like `read_config`, its return value depends on
    // the evaluation context (which cell owns the target).
    fn read_owner_config<'v>(
        section: StringValue,
        key: StringValue,
        default: Option<Value<'v>>,
        eval: &mut Evaluator<'v, '_, '_>,
    ) -> starlark::Result<Value<'v>> {
        // During `BUCK` file evaluation the build file cell is the owner cell, so
        // this is exactly `read_config`. We require a build file context, which means
        // this errors at the top level of a `.bzl` file (a `Bzl` context).
        if let Ok(build_context) = BuildContext::from_context(eval) {
            build_context.additional.require_build("read_owner_config")?;
            let buckconfigs = &build_context.buckconfigs;
            return match buckconfigs.current_cell_get(section, key, eval)? {
                Some(v) => Ok(v.to_value()),
                None => Ok(default.unwrap_or_else(Value::new_none)),
            };
        }

        // During analysis, read the buckconfig of the cell that owns the target.
        // This is pre-fetched and installed as `eval.extra` before analysis runs.
        if let Some(analysis_context) = AnalysisConfigContext::from_context(eval) {
            return match analysis_context.read_config(section.as_str(), key.as_str()) {
                Some(v) => Ok(eval.heap().alloc_str(&v).to_value()),
                None => Ok(default.unwrap_or_else(Value::new_none)),
            };
        }

        Err(buck2_error::Error::from(ReadOwnerConfigError::Unavailable).into())
    }

```
</details>

cormacrelf committed:
Make read_config work in bxl again

Previously I believe it only worked because read_config was speculative_exec_safe.
That is no longer the case. We make a late binding to read the BXL cell's config
more directly.

JakobDegen reviewed (approved):
Review automatically exported from Phabricator review in Meta.
